### PR TITLE
Fix NestedPool for py38

### PR DESCRIPTION
**Pull request by adrien-berchet**
### Context

NestedPool fails for Py38

### Resolution

Force the ``group`` argument to None in the constructor of ``NestedPool.Process``.

### CODE REVIEW

**Comment by adrien-berchet** [comment]
I had to update the CI to make it work both in GitHub and on my local machine.
But I don't understand why the py38 tests are not run? (tox just does nothing)

**Comment by adrien-berchet** [comment]
Any idea @asanin-epfl ?

File: .github/workflows/run-tox.yml [hunk 1]
```diff
@@ -6,7 +6,7 @@ on:
 jobs:
   build:
 
-    runs-on: ubuntu-latest
+    runs-on: ubuntu-18.04
```
Comment (asanin-epfl) [inline .github/workflows/run-tox.yml:9]: Why this change? Could it be that this version does not have py38 => tox does not run it?

File: morph_tool/nrnhines.py [hunk 2]
```diff
@@ -278,7 +278,19 @@ class NestedPool(multiprocessing.pool.Pool):  # pylint: disable=abstract-method
 
     class Process(multiprocessing.Process):
         """Class that represents a non-daemon process"""
-        daemon = False
+
+        def __init__(self, *args, **kwargs):
+            """Ensures group=None."""
+            kwargs.pop("group", None)
```
Comment (asanin-epfl) [inline morph_tool/nrnhines.py:284]:
```python
if 'group' in kwargs
```
looks clearer as it does not require a comment that explains this adhoc.

File: morph_tool/nrnhines.py [hunk 3]
```diff
@@ -278,7 +278,19 @@ class NestedPool(multiprocessing.pool.Pool):  # pylint: disable=abstract-method
...
+            super().__init__(None, *args[1:], **kwargs)
+
+        @property
```
Comment (asanin-epfl) [inline morph_tool/nrnhines.py:287]: A question. If `daemon` argument was passed to constructor, what would happen here?

File: setup.py [hunk 4]
```diff
@@ -16,7 +16,7 @@
     'click>=6.7',
     'deprecation>=2.1.0',
     'more-itertools>=8.6.0',
-    'morphio>=2.7.0',
+    'morphio>=2.7.0,<3',
     'neurom>=1.8.0',
```
Comment (asanin-epfl) [inline setup.py:20]: NeuroM v2 will require MorphIO > 3. That could conflict here in future.

File: tox.ini [hunk 5]
```diff
@@ -16,7 +16,7 @@ envlist =
 [testenv]
 extras = all
 deps = {[base]testdeps}
-commands = nosetests -sv tests
+commands = nosetests -sv tests {posargs}
```
Comment (asanin-epfl) [inline tox.ini:19]: Where does `posargs` come from? Is this change only for manual runs on a local machine?

File: setup.py (hunk 4 above)
Comment (adrien-berchet) [reply setup.py]: The test `test_diff.test_equality` fails with morphio==3. I can updated it to make it work with all versions.

File: tox.ini (hunk 5 above)
Comment (adrien-berchet) [reply tox.ini:19]: It is only for local runs. For example, if I want to debug with pdb, I have to pass the `-s` flag to nosetest, so thanks to posargs I can use the command `tox -e py36 -- -s`.

File: morph_tool/nrnhines.py (hunk 3 above)
Comment (adrien-berchet) [reply morph_tool/nrnhines.py:287]: No idea :-)
I think it would just try to set the `deamon` attribute to True, which would be intercepted by the setter.

File: morph_tool/nrnhines.py (hunk 2 above)
Comment (adrien-berchet) [reply morph_tool/nrnhines.py:284]: But is would perform 2 searches in the kwargs dict instead of one. Isn't it clear enough like this?

File: .github/workflows/run-tox.yml (hunk 1 above)
Comment (adrien-berchet) [reply .github/workflows/run-tox.yml:9]: That's what I wanted to check but no, tox still does do anything...

Comment (adrien-berchet) [reply .github/workflows/run-tox.yml]: Ah I found it: py38 was missing in tox.ini.

Comment (asanin-epfl) [reply .github/workflows/run-tox.yml]: good

File: morph_tool/nrnhines.py (hunk 2 above)
Comment (asanin-epfl) [reply morph_tool/nrnhines.py:284]: readability must come first

File: morph_tool/nrnhines.py (hunk 3 above)
Comment (asanin-epfl) [reply morph_tool/nrnhines.py:287]: this needs to be verified otherwise why having it

File: setup.py (hunk 4 above)
Comment (asanin-epfl) [reply setup.py]: Does morph-tool require NeuroM v2 or v1?

File: tox.ini (hunk 5 above)
Comment (asanin-epfl) [reply tox.ini:19]: ok, good to know

File: setup.py (hunk 4 above)
Comment (adrien-berchet) [reply setup.py]: I just tested with `NeuroM==2.0.1.dev8` and the test `test_nrnhines.test_NeuroM_section_to_NRN_compartment_paths` fails. But since NeuroM v2 is not released yet we could fix this in another PR?

File: morph_tool/nrnhines.py (hunk 2 above)
Comment (adrien-berchet) [reply morph_tool/nrnhines.py:284]: Well, I would say it depends :-)
But this method is not supposed to be called millions of times so let's change it.

File: morph_tool/nrnhines.py (hunk 3 above)
Comment (arnaudon) [reply morph_tool/nrnhines.py:287]: this stuff is a bit magical, and is used in several places at bbp (bluepymm, etc...). I wouldn't bother looking at these details tbh.

File: setup.py (hunk 4 above)
Comment (arnaudon) [reply setup.py]: yes, I'll make another PR to upgrade to morphio v3

**Review by arnaudon — APPROVED** [review]

File: morph_tool/nrnhines.py (hunk 2 above)
Comment (asanin-epfl) [reply morph_tool/nrnhines.py:284]: In the comment it's better to explain why `group` must be set to `None`

File: morph_tool/nrnhines.py (hunk 3 above)
Comment (adrien-berchet) [reply morph_tool/nrnhines.py:288]: I tried with the `daemon` argument and the test `test_nrnhines.test_isolate` fails. It seems that only this way works.

File: morph_tool/nrnhines.py (hunk 2 above)
Comment (adrien-berchet) [reply morph_tool/nrnhines.py]: I added a short explanation

File: morph_tool/nrnhines.py [hunk 6]
```diff
@@ -278,7 +278,25 @@ class NestedPool(multiprocessing.pool.Pool):  # pylint: disable=abstract-method
...
+        def __init__(self, *args, **kwargs):
+            """Ensures group=None to avoid exception from multiprocessing.
+
+            Note:
+                The exception raised by multiprocessing is:
+                AssertionError: group argument must be None for now
+            """
+            if "group" in kwargs:
+                kwargs.pop("group", None)
+            super().__init__(None, *args[1:], **kwargs)
```
Comment (asanin-epfl) [inline morph_tool/nrnhines.py:291]: I rather meant this
```suggestion
            if "group" in kwargs:
                kwargs.pop("group", None)
            else:
                args.pop(0)
            super().__init__(None, *args, **kwargs)
```
This way it is clear that `group` is the first argument without looking into documentation.

Comment (asanin-epfl) [reply morph_tool/nrnhines.py:291]: But now I'm thinking. If this is a repeated pattern then it's better keep it as it was. However it's a bad pattern IMO.

Comment (adrien-berchet) [reply morph_tool/nrnhines.py:291]: The `args` are often empty here, so the `args.pop(0)` will fail.
And having another `if` to check the length of args would be very heavy, for both readability and perf.

Comment (asanin-epfl) [reply morph_tool/nrnhines.py:291]: if `group` not in `kwargs` then it is in args, no? Anyway, lets keep the old code.

Comment (asanin-epfl) [reply morph_tool/nrnhines.py:291]: I mean this one
```
kwargs.pop("group", None)
super().__init__(None, *args[1:], **kwargs)
```

Comment (adrien-berchet) [reply morph_tool/nrnhines.py]: Ok, done :)

**Review by asanin-epfl — APPROVED** [review]
